addons/udes_stock/models/stock_quant_package.py

```diff
@@ -0,0 +1,112 @@
+# -*- coding: utf-8 -*-
+
+from odoo import api, fields, models, _
+from collections import defaultdict
+
+import logging
+
+_logger = logging.getLogger(__name__)
+
+
+class StockQuantPackage(models.Model):
+
+    _name = 'stock.quant.package'
+
+    _inherit = ['stock.quant.package', 'abstract.stock.model']
+
+    CREATE = True
+
+    # This works - but there must be a nicer way
+    name = fields.Char(
+        'Package Reference',
+        copy=False,
+        index=True,
+        default=lambda self: self.env['ir.sequence']
+        .next_by_code('stock.quant.package')
+        .replace('PACK', 'UDES')
+        or _('Unknown Pack'),
+    )
+
+    def _get_product_quantities_by_key(self, get_key=lambda q: q.product_id):
+        """ This function computes the product quantities the given package grouped by a key
+            :args:
+                get_key: a callable which takes a quant and returns the key
+        """
+        res = defaultdict(int)
+        for quant in self._get_contained_quants():
+            res[get_key(quant)] += quant.quantity
+        return res
+
+    def mls_can_fulfil(self, mls):
+        """ Returns mls which the package can fulfil. If the product_qty of the
+            mls is larger than in the package (i.e. in self) the mls will be split.
+        """
+        MoveLines = self.env['stock.move.line']
+        pack_quantities = self._get_all_products_quantities()
+        can_fulfil_mls = MoveLines.browse()
+        excess_mls = MoveLines.browse()
+        for prod, mls_grp in mls.groupby('product_id'):
+            pack_qty = pack_quantities.get(prod, 0)
+            if pack_qty == 0:
+                # just skip over
+                continue
+            fulfil_mls, excess_ml, _ = mls_grp.move_lines_for_qty(pack_qty)
+            can_fulfil_mls |= fulfil_mls
+            if excess_ml:
+                excess_mls |= excess_ml
+        return can_fulfil_mls, excess_mls
+
+    def has_same_content(self, other, get_values=lambda x: x.items()):
```

Comment:
What is the purpose of `get_values()` ?

Reply:
I did this when i was messing around with it, , just comparing the values and keys instead of the whole package- will remove the get_values behaviour